glib-macros/src/clone.rs

```diff
@@ -1,952 +1,495 @@
+            //
+            // If this is not an attribute then it is a closure, async closure or async block which
+            // is handled outside the loop
+            let attrs = input.call(Attribute::parse_outer)?;
+            let Some(attr) = attrs.first() else {
+                break;
+            };
+            let attr_name = attr.path();
+            if let Some(capture_kind) = attr_name
+                .get_ident()
+                .and_then(|i| CaptureKind::try_from(i).ok())
+            {
+                if capture_kind == CaptureKind::Watch {
+                    return Err(syn::Error::new_spanned(
+                        attr,
+                        "Watch variable captures are not supported",
+                    ));
                 }
-                None => {
-                    return Err(parts.generate_error_with_span(
-                        "Unexpected end after `as` keyword",
-                        current_span,
-                    ))
+                if attrs.len() > 1 {
+                    return Err(syn::Error::new_spanned(
+                        &attrs[1],
+                        "Variable capture attributes must be followed by an identifier",
+                    ));
+                }
+                captures.push(Capture::new(attr, input)?);
+            } else if attr_name.is_ident("default_return") {
+                if attrs.len() > 1 {
+                    return Err(syn::Error::new_spanned(
+                        &attrs[1],
+                        format!(
+                            "The `default_return` attribute must not be followed by any other attributes. Found {} more attribute{}",
+                            attrs.len() - 1,
+                            if attrs.len() > 2 { "s" } else { "" },
+                        )
+                    ));
                 }
-            }
-        }
-        Some(TokenTree::Ident(p)) => {
-            let err = format!("Unexpected `{p}`");
-            return Err(parts.generate_error(&err));
-        }
-        _ => None,
-    };
-    if name == "self" && alias.is_none() {
-        return Err(parts.generate_error_with_span(
-            "Can't use `self` as variable name. Try storing it in a temporary variable or \
-                rename it using `as`.",
-            name_span,
-        ));
-    } else if name.contains('.') && alias.is_none() {
-        let err = format!("`{name}`: Field accesses are not allowed as is, you must rename it!");
-        return Err(parts.generate_error_with_span(&err, name_span));
-    }
 
-    elements.push(ElemToClone {
-        name,
-        alias,
-        borrow_kind,
-    });
-    Ok(())
-}
+                default_return = DefaultReturn::new(attr, input)?;
+
+                let next_attrs = &input.call(Attribute::parse_outer)?;
+                if !next_attrs.is_empty() {
+                    return Err(syn::Error::new_spanned(
+                        &next_attrs[0],
+                        format!(
```

Comment:
It's definitely worth having a common function since it has the same text and the logic as the error above.

Reply:
Except that it's not :rofl: It's looking at different arrays, different indices, does the length handling different.

Reply:
I mean for the error, you can pass value, slice and everything.